### ISSUE TITLE
[BUG] Fast-failover lease needs a fail-safe mechanism to clear delinquent state.

### ISSUE THREAD

**Issue by james-munson** [longhorn/longhorn#9093]
## Describe the bug

As found in #9089, fast-failover handling relies on creation of a new share-manager pod to clear the lease's delinquent state.  If the creation is not possible, the state may not be cleared, leading to a deadlock if the delinquent state prevents remedial actions that would free up pod creation.

It would be good to have a fail-safe way to know to clear that state, regardless.   I think that might be when the holder node goes NotReady.  At that point, there is no use having special rules - the ordinary ones can apply.  It would be absolutely safe to clear any leases with a not-ready node as the holder. 

## To Reproduce

There is no repro, since @PhanLe1010 has covered the case in #9089.  This issue generalizes that situation.

## Expected behavior

"Delinquent" state of the lease should be cleared when holder node goes `NotReady`, regardless of volume's share-manager state.

## Support bundle for troubleshooting

N/A

## Environment

 - Longhorn version: 
 - Impacted volume (PV): 
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl):
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version:
   - Number of control plane nodes in the cluster:
   - Number of worker nodes in the cluster:
 - Node config
   - OS type and version:
   - Kernel version:
   - CPU per node:
   - Memory per node:
   - Disk type (e.g. SSD/NVMe/HDD):
   - Network bandwidth between the nodes (Gbps):
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal):
 - Number of Longhorn volumes in the cluster:

**Comment by james-munson** [comment]
In discussion with @PhanLe1010 and @ejweber, we view this as a nice-to-have, but not a must-have for 1.7.0.  Marking as 1.7.1.

**Comment by derekbit** [comment]
@james-munson If users unfortunately fall into the deadlock in v1.7.0, what can they do?

**Comment by derekbit** [comment]
Set the milestone to v1.8.0 and backport to 1.7.1. Nice to have in v1.7.0.

**Comment by james-munson** [comment]
The workaround is simply to delete the lease object.  That will clear the state, and a new lease will be created from scratch at the same time as the new share-manager pod.

**Comment by innobead** [comment]
> The workaround is simply to delete the lease object.  That will clear the state, and a new lease will be created from scratch at the same time as the new share-manager pod.

Can we make this in the important note as a know issue? Thanks.

**Comment by derekbit** [comment]
@james-munson What's the status of the ticket?

**Comment by james-munson** [comment]
Repro and test steps can be used from https://github.com/longhorn/longhorn-manager/pull/3013#issuecomment-2253358185

In addition, normal rwx-volume-fast-failover testing should see no change.

**Comment by derekbit** [comment]
@james-munson 
Sorry. I have one question.

> As found in https://github.com/longhorn/longhorn/issues/9089, fast-failover handling relies on creation of a new share-manager pod to clear the lease's delinquent state. If the creation is not possible, the state may not be cleared, leading to a deadlock if the delinquent state prevents remedial actions that would free up pod creation.

The step 6 in https://github.com/longhorn/longhorn/issues/9089#issue-2431167084 turns on the node of the share-manager pod, but why is the share manager pod creation still stuck?

In addition, can you elaborate more on `If the creation is not possible`? When will it happen? Any conditions?

**Comment by james-munson** [comment]
> The step 6 in [#9089 (comment)](https://github.com/longhorn/longhorn/issues/9089#issue-2431167084) turns on the node of the share-manager pod, but why is the share manager pod creation still stuck?

Good question.  One use of the "delinquent" status is to avoid trying to run or communicate with an engine or replica on the suspected node.  To do so will likely result in a timeout as the connection can't be made, and that will use up the fast failover time budget.  Ordinarily, that would not be a problem, but in the case in #9089 has the one usable replica for salvage on the suspected node.  If we can't talk to it, we can't bring it back online.

The deadlock situation is that the new share-manager pod can't mount its filesystem and get a running nfs-ganesha process, so it is hung in creation and cannot clear the "delinquent" state.  But the filesystem can't be mounted until the replica is salvaged, which is blocked by the "delinquent" state.

The fix for #9089 looks specifically for the volume in faulted state, and clears delinquency in that case.  I can't necessarily construct other cases where the ability to create the share-manager pod would be blocked by the moratorium on calls to the instance manager, but I would not be surprised if they exist.  Better not to find them by accident while testing other corner cases.  So the fix in #9093 just generalizes the idea to say that if the delinquent node is officially recognized as Not Ready, we might as well clear the delinquent status because it serves no purpose any more.  The time for fast failover has already come and gone, and we will use the regular failover handling.

In fact, with the fix for #9093, we could remove the specific fix for #9089, and that is how I tested it.

Does that explanation make sense?

**Comment by yangchiu** [comment]
Verification passed on master-head (longhorn-manager ) following [the test plan](https://github.com/longhorn/longhorn/issues/9093#issuecomment-2533222209).

1. Install Longhorn master-head on a 3 worker nodes
1. Enable RWX fast failover setting 
1. Deploy a workload which use a RWX PVC provisioned by Longhorn. For example, https://github.com/longhorn/longhorn/blob/master/examples/rwx/rwx-nginx-deployment.yaml
1. Reduce the number of replicas to 1. Delete 2 replicas not on the same node as the share manager pod
1. Turn off the node of the share manager pod
1. Volume stay in faulted, lease CR is not delinquent or stale. Share manager pod is recreated but then error and deleted. Cluster is stable, no error log storm. CPU usage is small

```
$ kubectl get lease -n longhorn-system
NAMESPACE         NAME                                             HOLDER                                                                      AGE
longhorn-system   driver-longhorn-io                               1734397429557-9345-driver-longhorn-io                                       44m
longhorn-system   external-attacher-leader-driver-longhorn-io      csi-attacher-8477ddf44c-tn6k6                                               44m
longhorn-system   external-resizer-driver-longhorn-io              csi-resizer-598f56f4f5-zw7sk                                                44m
longhorn-system   external-snapshotter-leader-driver-longhorn-io   csi-snapshotter-689d596cdb-h842g                                            44m
longhorn-system   longhorn-manager-upgrade-lock                                                                                                46m
longhorn-system   pvc-587e4625-45ca-4c96-bb77-7f163435a674         ip-10-0-1-165                                                               10m

$ kubectl get lease pvc-587e4625-45ca-4c96-bb77-7f163435a674 -n longhorn-system -oyaml
apiVersion: coordination.k8s.io/v1
kind: Lease
metadata:
  creationTimestamp: "2024-12-17T01:38:08Z"
  labels:
    longhorn.io/managed-by: longhorn-manager
    longhorn.io/share-manager: pvc-587e4625-45ca-4c96-bb77-7f163435a674
  name: pvc-587e4625-45ca-4c96-bb77-7f163435a674
  namespace: longhorn-system
  ownerReferences:
  - apiVersion: longhorn.io/v1beta2
    controller: false
    kind: ShareManager
    name: pvc-587e4625-45ca-4c96-bb77-7f163435a674
    uid: a6e7631b-2eb9-43ec-9c5a-fd458e311ab6
  resourceVersion: "5914"
  uid: 406656ef-ace8-4dd1-8edd-1e3c472c801c
spec:
  acquireTime: "0001-01-01T00:00:01.000000Z"
  holderIdentity: ip-10-0-1-165
  leaseDurationSeconds: 7
  leaseTransitions: 1
  renewTime: null
```

```
$ kubectl get pods -owide -w -n longhorn-system
NAME                                                     READY   STATUS    RESTARTS   AGE   IP           NODE            NOMINATED NODE   READINESS GATES
csi-attacher-8477ddf44c-6vjps                            1/1     Running   0          35m   10.42.1.13   ip-10-0-1-165   <none>           <none>
csi-attacher-8477ddf44c-mh8c4                            1/1     Running   0          35m   10.42.2.10   ip-10-0-1-216   <none>           <none>
csi-attacher-8477ddf44c-tn6k6                            1/1     Running   0          35m   10.42.3.10   ip-10-0-1-234   <none>           <none>
csi-provisioner-5847bdddf5-d7xbk                         1/1     Running   0          35m   10.42.1.14   ip-10-0-1-165   <none>           <none>
csi-provisioner-5847bdddf5-n9nbz                         1/1     Running   0          35m   10.42.3.12   ip-10-0-1-234   <none>           <none>
csi-provisioner-5847bdddf5-nwc4d                         1/1     Running   0          35m   10.42.2.11   ip-10-0-1-216   <none>           <none>
csi-resizer-598f56f4f5-jr577                             1/1     Running   0          35m   10.42.2.12   ip-10-0-1-216   <none>           <none>
csi-resizer-598f56f4f5-x9lhp                             1/1     Running   0          35m   10.42.1.16   ip-10-0-1-165   <none>           <none>
csi-resizer-598f56f4f5-zw7sk                             1/1     Running   0          35m   10.42.3.11   ip-10-0-1-234   <none>           <none>
csi-snapshotter-689d596cdb-7g824                         1/1     Running   0          35m   10.42.3.13   ip-10-0-1-234   <none>           <none>
csi-snapshotter-689d596cdb-h842g                         1/1     Running   0          35m   10.42.2.13   ip-10-0-1-216   <none>           <none>
csi-snapshotter-689d596cdb-jp4mt                         1/1     Running   0          35m   10.42.1.15   ip-10-0-1-165   <none>           <none>
engine-image-ei-b907910b-4khgn                           1/1     Running   0          36m   10.42.2.8    ip-10-0-1-216   <none>           <none>
engine-image-ei-b907910b-6jlvd                           1/1     Running   0          36m   10.42.3.8    ip-10-0-1-234   <none>           <none>
engine-image-ei-b907910b-kpwgx                           1/1     Running   0          36m   10.42.1.9    ip-10-0-1-165   <none>           <none>
instance-manager-4150b63fdd35f681e6690b230d549742        1/1     Running   0          35m   10.42.1.10   ip-10-0-1-165   <none>           <none>
instance-manager-5c4ffbe361eb94ec8eea2d1804aa6169        1/1     Running   0          35m   10.42.3.9    ip-10-0-1-234   <none>           <none>
instance-manager-f671f69d7cb1f1906103e4784537b474        1/1     Running   0          35m   10.42.2.9    ip-10-0-1-216   <none>           <none>
longhorn-csi-plugin-fgxmp                                3/3     Running   0          35m   10.42.2.14   ip-10-0-1-216   <none>           <none>
longhorn-csi-plugin-tqv8b                                3/3     Running   0          35m   10.42.1.17   ip-10-0-1-165   <none>           <none>
longhorn-csi-plugin-xclss                                3/3     Running   0          35m   10.42.3.14   ip-10-0-1-234   <none>           <none>
longhorn-driver-deployer-f5455fd6f-9zzq4                 1/1     Running   0          36m   10.42.2.6    ip-10-0-1-216   <none>           <none>
longhorn-manager-bngv8                                   2/2     Running   0          36m   10.42.1.8    ip-10-0-1-165   <none>           <none>
longhorn-manager-mhhd2                                   2/2     Running   0          36m   10.42.2.5    ip-10-0-1-216   <none>           <none>
longhorn-manager-rhxfd                                   2/2     Running   0          36m   10.42.3.6    ip-10-0-1-234   <none>           <none>
longhorn-ui-79dbd8bf77-ngkpr                             1/1     Running   0          36m   10.42.3.7    ip-10-0-1-234   <none>           <none>
longhorn-ui-79dbd8bf77-sstx7                             1/1     Running   0          36m   10.42.2.7    ip-10-0-1-216   <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   1/1     Running   0          32s   10.42.1.18   ip-10-0-1-165   <none>           <none>
longhorn-manager-bngv8                                   2/2     Running   0          37m   10.42.1.8    ip-10-0-1-165   <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   1/1     Terminating   0          78s   10.42.1.18   ip-10-0-1-165   <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   1/1     Terminating   0          78s   10.42.1.18   ip-10-0-1-165   <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   1/1     Terminating   0          78s   10.42.1.18   ip-10-0-1-165   <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   0/1     Pending       0          0s    <none>       <none>          <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   0/1     Pending       0          0s    <none>       ip-10-0-1-234   <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   0/1     ContainerCreating   0          0s    <none>       ip-10-0-1-234   <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   0/1     Running             0          2s    10.42.3.16   ip-10-0-1-234   <none>           <none>
engine-image-ei-b907910b-kpwgx                           1/1     Running             0          37m   10.42.1.9    ip-10-0-1-165   <none>           <none>
instance-manager-4150b63fdd35f681e6690b230d549742        1/1     Running             0          36m   10.42.1.10   ip-10-0-1-165   <none>           <none>
csi-resizer-598f56f4f5-x9lhp                             1/1     Running             0          36m   10.42.1.16   ip-10-0-1-165   <none>           <none>
longhorn-manager-bngv8                                   2/2     Running             0          37m   10.42.1.8    ip-10-0-1-165   <none>           <none>
csi-attacher-8477ddf44c-6vjps                            1/1     Running             0          36m   10.42.1.13   ip-10-0-1-165   <none>           <none>
longhorn-csi-plugin-tqv8b                                3/3     Running             0          36m   10.42.1.17   ip-10-0-1-165   <none>           <none>
csi-snapshotter-689d596cdb-jp4mt                         1/1     Running             0          36m   10.42.1.15   ip-10-0-1-165   <none>           <none>
csi-provisioner-5847bdddf5-d7xbk                         1/1     Running             0          36m   10.42.1.14   ip-10-0-1-165   <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   0/1     Terminating         0          42s   10.42.3.16   ip-10-0-1-234   <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   0/1     Completed           0          46s   10.42.3.16   ip-10-0-1-234   <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   0/1     Completed           0          47s   10.42.3.16   ip-10-0-1-234   <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   0/1     Completed           0          47s   10.42.3.16   ip-10-0-1-234   <none>           <none>
```

```
$ kubectl top pods -n longhorn-system
NAME                                                CPU(cores)   MEMORY(bytes)           
longhorn-manager-mhhd2                              5m           64Mi            
longhorn-manager-rhxfd                              15m          62Mi                      
```

1. Turn on the node of the share-manager pod
1. Volume is salvaged. Share manager pod is recreated and become running. Lease CR is updated normal renewal time and acquired time. Cluster is stable, no error log storm. CPU usage is small:

```
$ kubectl get pods -owide -w -n longhorn-system
NAME                                                     READY   STATUS    RESTARTS      AGE   IP           NODE            NOMINATED NODE   READINESS GATES
csi-attacher-8477ddf44c-7v7pk                            1/1     Running   0             28m   10.42.3.20   ip-10-0-1-234   <none>           <none>
csi-attacher-8477ddf44c-mh8c4                            1/1     Running   0             70m   10.42.2.10   ip-10-0-1-216   <none>           <none>
csi-attacher-8477ddf44c-tn6k6                            1/1     Running   0             70m   10.42.3.10   ip-10-0-1-234   <none>           <none>
csi-provisioner-5847bdddf5-jsxnm                         1/1     Running   0             28m   10.42.3.18   ip-10-0-1-234   <none>           <none>
csi-provisioner-5847bdddf5-n9nbz                         1/1     Running   0             70m   10.42.3.12   ip-10-0-1-234   <none>           <none>
csi-provisioner-5847bdddf5-nwc4d                         1/1     Running   0             70m   10.42.2.11   ip-10-0-1-216   <none>           <none>
csi-resizer-598f56f4f5-jr577                             1/1     Running   0             70m   10.42.2.12   ip-10-0-1-216   <none>           <none>
csi-resizer-598f56f4f5-sslsk                             1/1     Running   0             28m   10.42.2.19   ip-10-0-1-216   <none>           <none>
csi-resizer-598f56f4f5-zw7sk                             1/1     Running   0             70m   10.42.3.11   ip-10-0-1-234   <none>           <none>
csi-snapshotter-689d596cdb-7g824                         1/1     Running   0             70m   10.42.3.13   ip-10-0-1-234   <none>           <none>
csi-snapshotter-689d596cdb-h842g                         1/1     Running   0             70m   10.42.2.13   ip-10-0-1-216   <none>           <none>
csi-snapshotter-689d596cdb-sv26l                         1/1     Running   0             28m   10.42.2.18   ip-10-0-1-216   <none>           <none>
engine-image-ei-b907910b-4khgn                           1/1     Running   0             71m   10.42.2.8    ip-10-0-1-216   <none>           <none>
engine-image-ei-b907910b-6jlvd                           1/1     Running   0             71m   10.42.3.8    ip-10-0-1-234   <none>           <none>
engine-image-ei-b907910b-kpwgx                           1/1     Running   1 (21m ago)   71m   10.42.1.23   ip-10-0-1-165   <none>           <none>
instance-manager-4150b63fdd35f681e6690b230d549742        1/1     Running   0             21m   10.42.1.25   ip-10-0-1-165   <none>           <none>
instance-manager-5c4ffbe361eb94ec8eea2d1804aa6169        1/1     Running   0             70m   10.42.3.9    ip-10-0-1-234   <none>           <none>
instance-manager-f671f69d7cb1f1906103e4784537b474        1/1     Running   0             71m   10.42.2.9    ip-10-0-1-216   <none>           <none>
longhorn-csi-plugin-fgxmp                                3/3     Running   0             70m   10.42.2.14   ip-10-0-1-216   <none>           <none>
longhorn-csi-plugin-tqv8b                                3/3     Running   3 (21m ago)   70m   10.42.1.24   ip-10-0-1-165   <none>           <none>
longhorn-csi-plugin-xclss                                3/3     Running   0             70m   10.42.3.14   ip-10-0-1-234   <none>           <none>
longhorn-driver-deployer-f5455fd6f-9zzq4                 1/1     Running   0             71m   10.42.2.6    ip-10-0-1-216   <none>           <none>
longhorn-manager-bngv8                                   2/2     Running   2 (21m ago)   71m   10.42.1.22   ip-10-0-1-165   <none>           <none>
longhorn-manager-mhhd2                                   2/2     Running   0             71m   10.42.2.5    ip-10-0-1-216   <none>           <none>
longhorn-manager-rhxfd                                   2/2     Running   0             71m   10.42.3.6    ip-10-0-1-234   <none>           <none>
longhorn-ui-79dbd8bf77-ngkpr                             1/1     Running   0             71m   10.42.3.7    ip-10-0-1-234   <none>           <none>
longhorn-ui-79dbd8bf77-sstx7                             1/1     Running   0             71m   10.42.2.7    ip-10-0-1-216   <none>           <none>
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   1/1     Running   0             21m   10.42.1.21   ip-10-0-1-165   <none>           <none>
```

```
$ kubectl get lease pvc-587e4625-45ca-4c96-bb77-7f163435a674 -n longhorn-system -oyaml
apiVersion: coordination.k8s.io/v1
kind: Lease
metadata:
  creationTimestamp: "2024-12-17T01:38:08Z"
  labels:
    longhorn.io/managed-by: longhorn-manager
    longhorn.io/share-manager: pvc-587e4625-45ca-4c96-bb77-7f163435a674
  name: pvc-587e4625-45ca-4c96-bb77-7f163435a674
  namespace: longhorn-system
  ownerReferences:
  - apiVersion: longhorn.io/v1beta2
    controller: false
    kind: ShareManager
    name: pvc-587e4625-45ca-4c96-bb77-7f163435a674
    uid: a6e7631b-2eb9-43ec-9c5a-fd458e311ab6
  resourceVersion: "8123"
  uid: 406656ef-ace8-4dd1-8edd-1e3c472c801c
spec:
  acquireTime: "2024-12-17T01:53:15.305118Z"
  holderIdentity: ip-10-0-1-165
  leaseDurationSeconds: 7
  leaseTransitions: 2
  renewTime: "2024-12-17T01:55:39.312458Z"
```

```
$ kubectl top pods -n longhorn-system
NAME                                                     CPU(cores)   MEMORY(bytes)          
longhorn-driver-deployer-f5455fd6f-9zzq4                 1m           11Mi            
longhorn-manager-bngv8                                   15m          109Mi           
longhorn-manager-mhhd2                                   19m          64Mi            
longhorn-manager-rhxfd                                   19m          61Mi               
share-manager-pvc-587e4625-45ca-4c96-bb77-7f163435a674   5m           90Mi    
```
